### PR TITLE
Direct looking for AppendRelInfos by relid in the append_rel_array instead of trawl through the append_rel_list.

### PR DESCRIPTION
**Pull request by QingMa-gp**
Direct looking for ```AppendRelInfos``` by relid in the ```append_rel_array```
instead of trawl through the ```append_rel_list```.

Finding ```AppendRelInfos``` by relid had quite a large overhead when the
number of items in the ```append_rel_list``` was high, as it had to trawl
through the ```append_rel_list``` looking for ```AppendRelInfos``` belonging to
the given childrelids.

Upstream introduced ```append_rel_array``` to fix this performance issue
by storing an array in PlannerInfo which indexes ```AppendRelInfos``` by
child relid. Reduced time complexity from O(N) to O(1).

Discussion: https://postgr.es/m/94dd7a4b-5e50-0712-911d-2278e055c622@dalibo.com
commit: https://github.com/greenplum-db/gpdb/commit/7d872c91a3f9d49b56117557cdbb0c3d4c620687

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by kainwen** [comment]
> 

Typo in the commit messsage.

File: src/backend/cdb/cdbpathlocus.c [hunk 1]
```diff
@@ -432,19 +432,7 @@ cdbpathlocus_from_subquery(struct PlannerInfo *root,
...
-			{
-				AppendRelInfo *appendrel = lfirst(lc);
-
-				if (appendrel->child_relid == rel->relid)
-				{
-					parent_relid = appendrel->parent_relid;
-					break;
-				}
-			}
+			Index parent_relid = root->append_rel_array[rel->relid]->parent_relid;
```
Comment (kainwen) [inline src/backend/cdb/cdbpathlocus.c:435]: We need detailed comments here.

Also, should we check null pointer here?

File: src/backend/cdb/cdbpathlocus.c [hunk 2]
```diff
@@ -432,19 +432,7 @@ cdbpathlocus_from_subquery(struct PlannerInfo *root,
...
 
 			if (parent_relid <= 0)
```
Comment (kainwen) [inline src/backend/cdb/cdbpathlocus.c:437]: Do we need to change the logic here?

**Review by kainwen — COMMENTED** [review]
We need also add more words in commit message and justify that why not adding a case in this PR? Do we already have case cover this? If so, what case?

File: src/backend/cdb/cdbpathlocus.c (hunk 1 above)
Comment (QingMa-gp) [reply src/backend/cdb/cdbpathlocus.c]: Done.

File: src/backend/cdb/cdbpathlocus.c (hunk 2 above)
Comment (QingMa-gp) [reply src/backend/cdb/cdbpathlocus.c:443]: Added the reason into commit message.

**Comment by QingMa-gp** [comment]
The scenario has been covered by existing test case regress/partition_prune.

File: src/backend/cdb/cdbpathlocus.c [hunk 3]
```diff
@@ -432,26 +432,17 @@ cdbpathlocus_from_subquery(struct PlannerInfo *root,
 		 */
 		if (rel->reloptkind == RELOPT_OTHER_MEMBER_REL)
 		{
-			Index		parent_relid = 0;
+			Index parent_relid = 0;
```
Comment (kainwen) [inline src/backend/cdb/cdbpathlocus.c:435]: How about code like this?

```suggestion
			Index parent_relid;
			if (root->append_rel_array == NULL || root->append_rel_array[rel->relid] == NULL)
			    elog(ERROR, "cannot find valid append rel");
             parent_relid = root->append_rel_array[rel->relid];
```

Also remove the if-clause `if (parent_relid <= 0)`.

**Review by kainwen — APPROVED** [review]
One minor style issue:

1. one line code, we might skip the `{}`
2. pg c-style put `{` in a new line
3. we often put variable declartion at the beginning of a block.

Thanks for the PR.

File: src/backend/cdb/cdbpathlocus.c (hunk 3 above)
Comment (QingMa-gp) [reply src/backend/cdb/cdbpathlocus.c]: Good suggestion, edited.

**Comment by QingMa-gp** [comment]
> One minor style issue:
> 
> 1. one line code, we might skip the `{}`
> 2. pg c-style put `{` in a new line
> 3. we often put variable declartion at the beginning of a block.
> 
> Thanks for the PR.

Thanks for the review, already fixed the style issue.

**Pull request merged**